Unity-Technologies/ScriptableRenderPipeline
# [WIP] Refactor the shading framework (bis)

### Purpose of this PR

This PR is a replacement for https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/3250
The previous PR was hard to merge due to several change in both HDRP test and the code based + addition of DXR fail. Starting from fresh HDRP/Staging instead.

Refactor the shading framework to make it more easy to support arbitrary materials. Most of the non-material-specific code is now shared. The new CBSDF function which replaces the BSDF function is more flexible and allows to fully define the behavior of up to 4 types of lobes (more lobes can be grouped by type). The resulting code is efficient (automatic early-outs), simple and flexible.

Note: This PR is a draft and don't contain the change for StackLit 

Note2: I guess we have change the behavior of transmittance with directional is different, need to check that (i.e when there is transmission and no shadow with dir light)

Change in this PR:
- Introduce CBSDF, EvaluateBSDF, IsNonZeroBSDF to replace BSDF functions
- Update GetBSDFAngles parameters
- Update LightEvaluatoin and SurfaceShading functions, introduce define to enable/disable features of it
- Replace ComputeMicroShadowing by a function GetAmbientOcclusionForMicroShadowing() 
- Add shadow tint parameter on light
- Remove SunDisk and Sun Halo size
- Rename WorldToTangent to TangentToWorld as it was incorrectly named

---
### Release Notes
Refactor the shading framework to make it more easy to support arbitrary materials.

---
### Testing status
**Katana Tests**: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=HDRP%2Fshading-framework-refactor-bis&automation-tools_branch=master&unity_branch=trunk#
**Manual Tests**: WIP

What did you do?
- [X ] Opened test project + Run graphic tests locally
- [X ] C# and shader warnings (supress shader cache to see them)
- Other: 

**Automated Tests**: None.

Any test projects to go with this to help reviewers? No.

---
### Overall Product Risks
**Technical Risk**: High.

**Halo Effect**: High.

---
### Comments to reviewers

Compare to PR  https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/3250
- I use the name CBSDF instead of CxBSDF for consistency with all other naming (BSDFData, PostEvaluateBSDF, GetBSDFAngles etc...).
- I use IsNonZeroBSDF instead of IsNonZeroCBSDF
- I have replace ComputeMicroShadowing by a function GetAmbientOcclusionForMicroShadowing() that return ambient occlusion of BSDF data
- I fixed little mistake found here and there (like not the correct roughness clamp for hair, wrong name for diffuse variable for lambert in lit etc...)
- I added simple Lit support and stacklit
- I use define to avoid to have to add transmittance to AxF

@slunity I have done change for stacklit, you can update it if there is mistake. I think there is opportunity to clean some code here (code is quite complex currently), but I have try to keep same behavior.
